src/Util/EntityFieldQuery.php

```diff
@@ -0,0 +1,137 @@
+            $entity_table_alias,
+            $entity_info['entity keys']['id']
+          ));
+        }
+        elseif ($relational_filter->getType() == RelationalFilterInterface::TYPE_PROPERTY) {
+          // We only know about the uid in this scenario.
+          if ($relational_filter->getName() == 'uid') {
+            $user_table_alias = $this::aliasJoinTable('users', $select_query);
+            $select_query->addJoin('INNER', 'users', $user_table_alias, sprintf('%s.uid = %s.uid',
+              $entity_table,
+              $user_table_alias
+            ));
+          }
+          throw new ServerConfigurationException(sprintf('Unsupported nested filter on property %s', $relational_filter->getName()));
+        }
+      }
+      /* @var RelationalFilterInterface $condition */
+      if ($condition->getType() == RelationalFilterInterface::TYPE_FIELD) {
+        // Make the join to the filed table for the condition.
+        $field_table_name = _field_sql_storage_tablename(field_info_field($condition->getName()));
+        $field_column = _field_sql_storage_columnname($condition->getName(), $condition->getColumn());
+        $field_table_alias = $this::aliasJoinTable($field_table_name, $select_query);
+        $select_query->addJoin('INNER', $field_table_name, $field_table_alias, sprintf('%s.%s = %s.%s',
+          $entity_table_alias,
+          $entity_info['entity keys']['id'],
+          $field_table_alias,
+          $id_key
+        ));
+        if (in_array($relationship['operator'], array('IN', 'BETWEEN'))) {
+          $select_query->condition($field_table_name . '.' . $field_column, $relationship['value'], $relationship['operator'][0]);
+        }
+        else {
+          for ($index = 0; $index < count($relationship['value']); $index++) {
+            $select_query->condition($field_table_name . '.' . $field_column, $relationship['value'][$index], $relationship['operator'][$index]);
+          }
+        }
+      }
+      elseif ($condition->getType() == RelationalFilterInterface::TYPE_PROPERTY) {
+        if (in_array($relationship['operator'], array('IN', 'BETWEEN'))) {
+          $select_query->condition($entity_info['base table'] . '.' . $condition->getName(), $relationship['value'], $relationship['operator'][0]);
+        }
+        else {
+          for ($index = 0; $index < count($relationship['value']); $index++) {
+            $select_query->condition($entity_info['base table'] . '.' . $condition->getName(), $relationship['value'][$index], $relationship['operator'][$index]);
+          }
+        }
+      }
+    }
+    return parent::finishQuery($select_query, $id_key);
+  }
+
+  /**
+   * Helper function tha checks if the select query already has a join.
+   *
+   * @param string $table_name
+   *   The name of the table.
+   * @param \SelectQuery $query
+   *   The query.
+   *
+   * @return string
+   *   The table alias.
+   */
+  protected static function aliasJoinTable($table_name, \SelectQuery $query) {
```

Comment:
I think you can check it with the join type like [here](https://github.com/Gizra/og/blob/7.x-2.x/og.module#L1389)